drivers/gles2/rasterizer_storage_gles2.cpp

```diff
@@ -3388,6 +3388,24 @@ RasterizerStorage::MMInterpolator *RasterizerStorageGLES2::_multimesh_get_interp
 	return &multimesh->interpolator;
 }
 
+void RasterizerStorageGLES2::multimesh_attach_canvas_item(RID p_multimesh, RID p_canvas_item, bool p_attach) {
+	MultiMesh *multimesh = multimesh_owner.getornull(p_multimesh);
+	ERR_FAIL_NULL(multimesh);
+	ERR_FAIL_COND(!p_canvas_item.is_valid());
+
+	if (p_attach) {
+		int64_t found = multimesh->linked_canvas_items.find(p_canvas_item);
+		if (found == -1) {
+			multimesh->linked_canvas_items.push_back(p_canvas_item);
+		}
+	} else {
+		int64_t found = multimesh->linked_canvas_items.find(p_canvas_item);
+		if (found != -1) {
+			multimesh->linked_canvas_items.remove_unordered(found);
+		}
+	}
```

Comment:
Why the difference compared to the skeleton one? Shouldn't these be pretty much the same? :thinking:
https://github.com/godotengine/godot/blob/b776cf54982e5f5bcbe33c226eb984f8ac0ef3ed/drivers/gles2/rasterizer_storage_gles2.cpp#L3782-L3788

Reply:
The skeleton one is simpler because there can only be one skeleton linked by a `canvas_item` (it is stored as a `RID` on the `canvas_item`).

For `MultiMesh`, the situation is more complex, each `Item::Command` which stores a draw instruction on the `canvas_item` can potentially hold a link to a `MultiMesh`. Thus the same `MultiMesh` can contain _multiple_ links per `canvas_item`.

These links need to be simplified to build up and destroy such that when the commands referencing a particular `MultiMesh` are gone, the link is broken.

One obvious way would be to use a refcount. But in fact the situation here can be slightly simplified, because `canvas_items` never delete `Commands` once they are added. They are only ever cleared at once. So we can guarantee during deletion if any `Command` contains a link to a `MultiMesh`, we can destroy the link. Any later duplicate links will not find the `linked_item` (hence the test for `found`) in the destroy.

For creation it's the same deal. We only need to add the link on the first `Command` which references a `MultiMesh`. Later commands will find the already existing link and do nothing.

Reply:
Sorry, my explanation with the ratios was perhaps a little confusing, as there are lots of permutations.

The important point is that if you look at the definition of `Item` (rasterizer.h, line 801) you will see that only ONE skeleton can be used by an `Item`. Therefore no similar check is needed for skeleton, as the canvas item can only be linked once.

For multimeshes, there can be MULTIPLE multimeshes linked per `Item`.

Reply:
I mean yes, many errors theoretically should not occur and thus could be dev-only. Theoretically though... :wink:

To be clear: I'm not advocating for adding error checks everywhere. It's strange that both "should-never-happen" cases are treated differently, that's all. And I get the `ERR_FAIL_COND(found == -1);` has kinda no cost as `found` is available anyway but it's still kinda pointless as `LocalVector::remove_unordered` checks the index validity itself (and double checking because it's cheap doesn't seem reasonable). So alternative to adding more checks like in #80325 could be removing the existing `ERR_FAIL_COND(found == -1);` check. Or could be left as is, it's not a big thing anyway, just catched my eye.

Reply:
Correct about `remove_unordered()`, but optimizer should sort this kind of redundant thing. If there's a check for -1, and a later similar check (inlined), it should remove the latter. It's also imo a bit of a micro-optimization (unless it shows up in profiles), and on the plus side the error message is better this way. :grinning: 

The potentially expensive bit is the `find()` in worst cases (say the skeleton has 10,000 linked canvas items), then detaching each in turn will be log performance drop with `n`. This is the same problem we had with notifications.

That isn't to say there are other techniques / containers that avoid using `find()`, but I don't think it is warranted here.